### PR TITLE
Replace RwLock by a futex based one on Linux.

### PR DESCRIPTION
**Pull request by m-ou-se**
This replaces the pthread-based RwLock on Linux by a futex based one.

This implementation is similar to [the algorithm](https://gist.github.com/kprotty/3042436aa55620d8ebcddf2bf25668bc) suggested by @kprotty, but modified to prefer writers and spin before sleeping. It uses two futexes: One for the readers to wait on, and one for the writers to wait on. The readers futex contains the state of the RwLock: The number of readers, a bit indicating whether writers are waiting, and a bit indicating whether readers are waiting. The writers futex is used as a simple condition variable and its contents are meaningless; it just needs to be changed on every notification.

Using two futexes rather than one has the obvious advantage of allowing a separate queue for readers and writers, but it also means we avoid the problem a single-futex RwLock would have of making it hard for a writer to go to sleep while the number of readers is rapidly changing up and down, as the writers futex is only changed when we actually want to wake up a writer.

It always prefers writers, as we decided [here](https://github.com/rust-lang/rust/issues/93740#issuecomment-1070696128).

It relies on futex_wake to return the number of awoken threads to be able to handle write-unlocking while both the readers-waiting and writers-waiting bits are set. Instead of waking both and letting them race, it first wakes writers and only continues to wake the readers too if futex_wake reported there were no writers to wake up.

r? @Amanieu

### CODE REVIEW

**Comment by m-ou-se** [comment]
```
test process::tests::test_finish_twice has been running for over 60 seconds
test process::tests::test_interior_nul_in_env_key_is_error has been running for over 60 seconds
test process::tests::test_interior_nul_in_env_value_is_error has been running for over 60 seconds
test process::tests::test_override_env has been running for over 60 seconds
test sync::rwlock::tests::frob has been running for over 60 seconds
```
Looks like something is wrong. :)

**Comment by m-ou-se** [comment]
Ah, looks like I somehow forgot to ever _set_ the READERS_WAITING bit. Well that's an easy fix. :)

**Comment by m-ou-se** [comment]
Going to do some more testing first. :)

File: library/std/src/sys/unix/locks/futex_rwlock.rs [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+    pub unsafe fn read(&self) {
+        if !self.try_read() {
+            self.read_contended();
+        }
+    }
+
+    #[inline]
+    pub unsafe fn try_write(&self) -> bool {
+        self.state
+            .fetch_update(Acquire, Relaxed, |s| (readers(s) == 0).then(|| s + WRITE_LOCKED))
```
Comment (Amanieu) [inline library/std/src/sys/unix/locks/futex_rwlock.rs:70]: I don't think this correctly handles the case where there is an existing write lock but no waiting threads or readers.

File: library/std/src/sys/unix/locks/futex_rwlock.rs [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+        Self { state: AtomicI32::new(0), writer_notify: AtomicI32::new(0) }
+    }
+
+    #[inline]
+    pub unsafe fn destroy(&self) {}
+
+    #[inline]
+    pub unsafe fn try_read(&self) -> bool {
+        self.state
+            .fetch_update(Acquire, Relaxed, |s| read_lockable(s).then(|| s + READ_LOCKED))
```
Comment (Amanieu) [inline library/std/src/sys/unix/locks/futex_rwlock.rs:56]: What happens if the reader count overflows?

Comment (kprotty) [reply library/std/src/sys/unix/locks/futex_rwlock.rs:56]: `read_lockable()` checks for `MAX_READERS` which looks like its writer value minus one. If the reader count would overflow into a writer, it looks like that would be detected by `read_lockable()` and short circuit to `None`.

**Comment by m-ou-se** [comment]
(Found a few more issues, so this is back to the drawing board for now. Will send a better tested PR soon. :) )

**Comment by m-ou-se** [comment]
Sorry for the noise :)

Here's the new PR, this time properly stress tested: https://github.com/rust-lang/rust/pull/95801

**Pull request closed without merging**